### ISSUE TITLE
refactor Array handling in oa-graph

### ISSUE THREAD

**Issue by ndushay** [sul-dlss/oa-graph#7]
Try replacing with xxx.collect {|s| s.body.uri}. This pattern may apply to many sections of your code, so you might want to play with it first to be sure you're getting what you want and get familiar or comfortable with it.

See discussions of this ruby method, e.g.
http://stackoverflow.com/questions/5347949/whats-different-between-each-and-collect-method-in-ruby

Also, review the many ruby array convenience methods at http://ruby-doc.org/core-2.2.0/Array.html, such as .sample, .shuffle, .map (.collect), .delete, .delete_if, etc.

I also like this video resource that we have access to at lynda.com, see the 'Array methods' topic in chapter 2 and all the nifty stuff in chapter 4 at http://www.lynda.com/Ruby-tutorials/essential-training/47905-2.html

**Comment by ndushay** [comment]
Also, for statements_from_ttl_minus_fedora

"The last concern here is the logic around if g, consider:

```
[1] pry(main)> g = RDF::Graph.new
=> #<RDF::Graph:0x3fbc601af19c(default)>
[2] pry(main)> puts 'OK' if g
OK
=> nil
[3] pry(main)> puts 'OK' unless g.empty?
=> nil
```

Actually, maybe the method argument could be revised to something like:
```ruby
def statements_from_ttl_minus_fedora(ttl=nil)
  if ttl
     etc.etc
  end
end
```
"

**Comment by ndushay** [comment]
I have examined the code and all of the instances of .each  in oa-graph (in lib and spec both) really are about applying a block to _all_ the members of a collection, and not about mapping a collection to a new collection.  So I am closing this ticket.

**Comment by dazza-codes** [comment]
Let's discuss.

**Comment by dazza-codes** [comment]
``` ruby
oa_jsonld = '{"@context":"http://iiif.io/api/presentation/2/context.json","@graph":[{"@id":"_:g70349699654640","@type":["dctypes:Text","cnt:ContentAsText"],"chars":"I love this!","format":"text/plain","language":"en"},{"@type":"oa:Annotation","motivation":"oa:commenting","on":"http://purl.stanford.edu/kq131cs7229","resource":"_:g70349699654640"}]}'
g = RDF::Graph.new
format = RDF::Format.for(:jsonld)
format.reader.new(oa_jsonld) {|r| r.each_statement {|s| g << s }}
q = [:s, RDF::Vocab::OA.motivatedBy, :o]
g.query(q).each_object.collect {|o| o.to_s }
#=> ["http://www.w3.org/ns/oa#commenting"]
q = [:s, RDF::Vocab::OA.hasSource, :o]
g.query(q).each_object.collect {|o| o.to_s }
#=> []
# So note how it returns an empty array when it doesn't match anything.
# That's why there is no need to declare the array up front and then 
# populate it in the each block.  That's why there is no need for an `if`
# block, the RDF.rb and ruby array methods take care of all that.
```

**Comment by ndushay** [comment]
So I think I've applied your suggestion in the 3 places where it made sense;  it did tidy up the code a lot.  It had a slight effect on a spec:  when it was expecting a URI and didn't find one, it returned [nil] instead of [], but that could be addressed with compact.  :-)

There is one other place where it's possible your suggestion could apply, but i think it makes the code less clear rather than more clear.  https://github.com/sul-dlss/oa-graph/blob/master/lib/oa/graph.rb#L25-32

**Comment by ndushay** [comment]
- [x] cut a new release of the gem once PR is approved
- [x] update sul-dlss/triannon gem to use new oa-graph gem.

**Comment by dazza-codes** [comment]
RE:  https://github.com/sul-dlss/oa-graph/blob/master/lib/oa/graph.rb#L25-32

Yes, that one's OK because of the recursive call to expand the objects and the later use of `array.uniq` to collapse the result.  My only concern is whether this should be a different data structure entirely, i.e. an RDF::Graph instead of an array, e.g.

``` ruby
g = RDF::Graph.new
# then use g and return it instead of an array
graph.query([subject, nil, nil]).each { |stmt|
        g << stmt
        subject_statements(stmt.object, graph).each { |s| g << s }
      }
g
```

**Comment by ndushay** [comment]
I'm not keen to change the method signature;  the idea is that you can add the returned statements to ANOTHER graph that you're building, or remove them from an existing graph.  

It's certainly possible my design could be improved;  if you feel it's important please open a separate, new issue in oa_graph for it.

I'm going to close this one.

You seem to like very atomic PRs;  I tend to like atomic issues.  :smile:
